README.md
![BQuest Logo](https://raw.githubusercontent.com/ottogroup/bquest/main/docs/assets/logo.svg)

# BQuest

Effortlessly validate and test your Google BigQuery queries with the power of pandas DataFrames in Python.

We would like to thank [Mike Czech](https://github.com/mikeczech) who is the original inventor of bquest!

**Warning**

This library is a work in progress!

Breaking changes should be expected until a 1.0 release, so version pinning is recommended.

[![CI: Overall outcome](https://github.com/ottogroup/bquest/workflows/Tests/badge.svg)](https://github.com/ottogroup/bquest/actions?workflow=Tests)
[![CD: gh-pages documentation](https://github.com/ottogroup/bquest/actions/workflows/pages/pages-build-deployment/badge.svg?branch=gh-pages)](https://github.com/ottogroup/bquest/actions/workflows/pages/pages-build-deployment)
[![PyPI version](https://img.shields.io/pypi/v/bquest.svg)](https://pypi.org/project/bquest/)
[![Project status (alpha, beta, stable)](https://img.shields.io/pypi/status/bquest.svg)](https://pypi.python.org/pypi/bquest/)
[![PyPI downloads](https://static.pepy.tech/personalized-badge/bquest?period=month&units=international_system&left_color=grey&right_color=blue&left_text=PyPI%20downloads/month)](https://pepy.tech/project/bquest)
[![Project license](https://img.shields.io/github/license/ottogroup/bquest)](https://github.com/ottogroup/bquest/blob/main/LICENSE)
[![Python version compatibility](https://img.shields.io/pypi/pyversions/bquest.svg)](https://pypi.python.org/pypi/bquest/)
[![Documentation: Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## Overview

* Use BQuest in combination with your favorite testing framework (e.g. pytest).
* Create temporary test tables from [JSON](https://cloud.google.com/bigquery/docs/loading-data) or
[pandas.DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html).
* Run BQ configurations and plain SQL queries on your test tables and check the result.

## Installation

Via PyPi (standard):

```bash
    pip install bquest
```

Via Github (most recent):

```bash
    pip install git+https://github.com/ottogroup/bquest
```

BQuest also requires a dedicated BigQuery dataset for storing test tables, e.g.

```yaml
    resource "google_bigquery_dataset" "bquest" {
      dataset_id    = "bquest"
      friendly_name = "bquest"
      description   = "Source tables for bquest tests"
      location      = "EU"
      default_table_expiration_ms = 3600000
    }
```

We recommend setting an [expiration time](https://www.terraform.io/docs/providers/google/r/bigquery_dataset.html#default_table_expiration_ms) for tables in the bquest dataset to assure removal of those test tables upon
test execution.

## Example

Given a ``pandas.DataFrame``

| foo    | weight | prediction_date   |
|--------|--------|-------------------|
| bar    | 23     | 20190301          |
| my     | 42     | 20190301          |

and its table definition

```python
    from bquest.tables import BQTableDefinitionBuilder

    table_def_builder = BQTableDefinitionBuilder(GOOGLE_PROJECT_ID, dataset="bquest", location="EU")
    table_definition = table_def_builder.from_df("abc.feed_latest", df)
```

you can use the config file ``*./abc/config.py*``

```json-object
    {
        "query": """
            SELECT
                foo,
                PARSE_DATE('%Y%m%d', prediction_date)
            FROM
                `{source_table}`
            WHERE
                weight > {THRESHOLD}
        """,
        "start_date": "prediction_date",
        "end_date": "prediction_date",
        "source_tables": {"source_table": "abc.feed_latest"},
        "feature_table_name": "abc.myid",
    }
```

and the runner

```python
    from bquest.runner import BQConfigFileRunner, BQConfigRunner

    runner = BQConfigFileRunner(
        BQConfigRunner(bq_client, bq_executor_func),
        "config/bq_config",
    )

    result_df = runner.run_config(
        "20190301",
        "20190308",
        [table_definition],
        "abc/config.py",
        templating_vars={"THRESHOLD": "30"},
    )
```

to assert the result table

```python
    assert result_df.shape == (1, 2)
    assert result_df.iloc[0]["foo"] == "my"
```

## Testing

For the actual testing bquest relies on an accessible BigQuery project which can be configured
with the [gcloud](https://cloud.google.com/sdk/docs/install?hl=de) client. The corresponding 
``GOOGLE_PROJECT_ID`` is extracted from this project
and used with [pandas-gbq](https://github.com/googleapis/python-bigquery-pandas) to write temporary tables to the bquest dataset that has to be pre-
configured before testing on that project.

For Github CI we have configured an identity provider in our testing project which allows
only core members of this repository to access the testing projects' resources.

## Important Links

- Full documentation: https://ottogroup.github.io/bquest/


CHANGELOG.md
### Development

- Add test coverage report
- Make consistent use of original_table_id/test_table_id
- Add tests for is_sql

### 0.2.0 (2023-06-12)

- Remove redundant project_id in BQConfigRunner and SQLRunner

### 0.1.0 (2023-06-12)

- Lower pandas requirement to 1.5.0

### 0.0.1 (2023-06-09)

- Initial release

bquest/runner.py
"""Module for Running BQuest Tests"""
import ast
import copy
import os
from typing import Any, Callable, Dict, List, Optional

import pandas
from google.cloud import bigquery as bq

from bquest.tables import BQTable, BQTableDefinition, BQTableDefinitionBuilder


class BQConfigSubstitutor:
    """Substitutes parameters inside a BQ configuration"""

    def __init__(self, config: Dict[str, Any], allow_partial: Optional[bool] = False):
        self._config = config
        self._allow_partial = allow_partial

    @property
    def original_feature_table_name(self) -> str:
        return str(self._config["feature_table_name"])

    def _map_source_table_ids_to_mock_table_ids(self, source_tables: List[BQTable]) -> Dict[str, str]:
        """Match source tables with their mocks."""
        result = {}
        table_id_mapping = {t.original_table_id: t.fq_test_table_id for t in source_tables}

        # e.g. { "feed": "bquest.example_id" }
        for table_key, table_id in self._config["source_tables"].items():
            if table_id in table_id_mapping:
                result[table_key] = table_id_mapping[table_id]
            else:
                if not self._allow_partial:
                    raise ValueError(f"Found no substitution for table {table_id}")
                result[table_key] = table_id
        return result

    def substitute(
        self,
        start_date: str,
        end_date: str,
        feature_table_name: BQTable,
        test_tables: List[BQTable],
    ) -> Dict[str, Any]:
        """Substitutes a wide array of parameters inside a BQ configuration.

        Args:
            start_date: the start date
            end_date: the end date
            feature_table_name: the test table where the results will be stored
            test_tables: test tables that replace the original source tables

        Returns:
            Dict: a new BQ configuration where parameters have been substituted.
        """
        config = copy.deepcopy(self._config)
        config["start_date"] = start_date
        config["end_date"] = end_date
        config["feature_table_name"] = feature_table_name.fq_test_table_id
        config["source_tables"] = self._map_source_table_ids_to_mock_table_ids(test_tables)
        return config


class BaseRunner:
    """Base class for runners"""

    def __init__(self, bq_client: bq.Client, dataset: str = "bquest"):
        self._bq_client = bq_client
        self._bq_table_def_builder = BQTableDefinitionBuilder(bq_client.project, dataset)

    def _create_source_tables(self, table_definitions: List[BQTableDefinition]) -> List[BQTable]:
        result = []
        for table_def in table_definitions:
            test_table = table_def.load_to_bq(self._bq_client)
            result.append(test_table)
        return result

    def _create_result_table_from_def(self, table_definition: BQTableDefinition) -> BQTable:
        return table_definition.load_to_bq(self._bq_client)

    def _create_empty_result_table(self, table_name: str) -> BQTable:
        result_table_def = self._bq_table_def_builder.create_empty(table_name)
        return result_table_def.load_to_bq(self._bq_client)


class BQConfigRunner(BaseRunner):
    """Runs BQ configurations on custom data for testing"""

    def __init__(
        self,
        bq_client: bq.Client,
        bq_executor_func: Callable[[Dict[str, Any], Optional[Dict[str, str]]], None],
        dataset: str = "bquest",
        clean_up: bool = True,
    ):
        super().__init__(bq_client, dataset)
        self._bq_executor_func = bq_executor_func
        self._clean_up = clean_up

    def run_config(
        self,
        start_date: str,
        end_date: str,
        source_table_definitions: List[BQTableDefinition],
        substitutor: BQConfigSubstitutor,
        result_table_definition: Optional[BQTableDefinition] = None,
        templating_vars: Optional[Dict[str, str]] = None,
    ) -> pandas.DataFrame:
        """Runs a BQ configuration with custom table definitions.

        Args:
            start_date: the start date (e.g. 20190301)
            end_date: the end date (e.g. 20190308)
            source_table_definitions: custom table definitions that replace the source tables of the BQ configuration
            substitutor:  a substitutor for BQ configurations
            result_table_definition: optional result table definition used for creating an empty result table
            templating_vars: variables that are inserted into the given bq configuration
        Returns:
            the contents of the results table
        """
        source_tables = self._create_source_tables(source_table_definitions)
        result_table = (
            self._create_result_table_from_def(result_table_definition)
            if result_table_definition
            else self._create_empty_result_table(substitutor.original_feature_table_name)
        )

        test_bq_config = substitutor.substitute(start_date, end_date, result_table, source_tables)

        # run config with substituted table identifiers
        self._bq_executor_func(test_bq_config, templating_vars)

        return result_table.to_df()


class BQConfigFileRunner:
    """Class for Running BQConfigs"""

    def __init__(self, bq_config_runner: BQConfigRunner, config_base_path: str) -> None:
        self._bq_config_runner = bq_config_runner
        self._config_base_path = config_base_path

    def run_config(
        self,
        start_date: str,
        end_date: str,
        source_table_definitions: List[BQTableDefinition],
        path_to_config: str,
        result_table_definition: Optional[BQTableDefinition] = None,
        allow_partial_table_substitutions: Optional[bool] = False,
        templating_vars: Optional[Dict[str, str]] = None,
    ) -> pandas.DataFrame:
        """Runs a BQ configuration file"""
        with open(os.path.join(self._config_base_path, path_to_config), "r", encoding="UTF-8") as f:
            try:
                config = ast.literal_eval(f.read())
            except ValueError as e:
                raise ValueError("Could not read the configuration.") from e
            return self._bq_config_runner.run_config(
                start_date,
                end_date,
                source_table_definitions,
                BQConfigSubstitutor(config, allow_partial=allow_partial_table_substitutions),
                result_table_definition=result_table_definition,
                templating_vars=templating_vars,
            )


class SQLRunner(BaseRunner):
    """Runs SQL queries on custom data for testing"""

    def __init__(
        self,
        bq_client: bq.Client,
        dataset: str = "bquest",
        clean_up: Optional[bool] = True,
    ):
        """

        Args:
            bq_client: BigQuery client used for interaction with BigQuery
            dataset: dataset which will be used for testing
            clean_up:  boolean if tables should be cleaned up
        """
        super(SQLRunner, self).__init__(bq_client, dataset)
        self._bq_client = bq_client
        self._clean_up = clean_up

    def run(
        self,
        sql: str,
        source_table_definitions: List[BQTableDefinition],
        substitutions: Optional[Dict[str, str]] = None,
        string_replacements: Optional[Dict[str, str]] = None,
        result_table_definition: Optional[BQTableDefinition] = None,
    ) -> pandas.DataFrame:
        """

        Args:
            sql: SQL query that is being executed in BigQuery
            source_table_definitions: source table definitions, list of BQTableDefinition
            substitutions: substitutions for the given query
            string_replacements: entire string replacements for the query, substitutions are placed before
            result_table_definition: result table definition

        Returns:
            pandas DataFrame of result table
        """
        if substitutions is None:
            substitutions = {}

        if string_replacements is None:
            string_replacements = {}

        _ = self._create_source_tables(source_table_definitions)
        _ = (
            self._create_result_table_from_def(result_table_definition)
            if result_table_definition
            else self._create_empty_result_table("result")
        )

        sql_with_substitutions = sql.format(**substitutions)
        for key, value in string_replacements.items():
            sql_with_substitutions = sql_with_substitutions.replace(key, value)

        job_config = bq.QueryJobConfig()
        query_job = self._bq_client.query(sql_with_substitutions, job_config=job_config)
        query_job.result()

        return query_job.result().to_dataframe()


class SQLFileRunner:
    """Class for running SQLFiles."""

    def __init__(self, sql_runner: SQLRunner, base_path: str) -> None:
        """

        Args:
            sql_runner: SQLRunner used to execute SQL config
            base_path: base path where SQL config is looked for
        """
        self._sql_runner = sql_runner
        self._base_path = base_path

    def run(
        self,
        path_to_sql: str,
        source_table_definitions: List[BQTableDefinition],
        substitutions: Optional[Dict[str, str]] = None,
        string_replacements: Optional[Dict[str, str]] = None,
    ) -> pandas.DataFrame:
        """

        Args:
            path_to_sql: relative path to SQL config file based on base path
            source_table_definitions: source table definitions
            substitutions: substitutions for SQL query
            string_replacements: entire string replacements for the query, substitutions are placed before

        Returns:

        """
        if substitutions is None:
            substitutions = {}

        if string_replacements is None:
            string_replacements = {}

        file = os.path.join(self._base_path, path_to_sql)
        try:
            with open(file, "r", encoding="UTF-8") as f:
                sql = f.read()
        except IOError as e:
            raise ValueError(f"Could not read the SQL file {file}.") from e
        return self._sql_runner.run(sql, source_table_definitions, substitutions, string_replacements)
